# Python格式化字符串f-string概览

## For example
`print(f"Iteration {i:4}: Cost {J_history[-1]:0.2e} ",
       f"dj_dw: {dj_dw: 0.3e}, dj_db: {dj_db: 0.3e} ",
       f"w: {w: 0.3e}, b:{b: 0.5e}")`

In [1]:
def gradient_descent(x, y, w_in, b_in, alpha, num_iters, cost_function, gradient_function): 
    """
    Performs gradient descent to fit w,b. Updates w,b by taking 
    num_iters gradient steps with learning rate alpha
    
    Args:
      x (ndarray (m,))  : Data, m examples 
      y (ndarray (m,))  : target values
      w_in,b_in (scalar): initial values of model parameters  
      alpha (float):     Learning rate
      num_iters (int):   number of iterations to run gradient descent
      cost_function:     function to call to produce cost
      gradient_function: function to call to produce gradient
      
    Returns:
      w (scalar): Updated value of parameter after running gradient descent
      b (scalar): Updated value of parameter after running gradient descent
      J_history (List): History of cost values
      p_history (list): History of parameters [w,b] 
      """
    
    w = copy.deepcopy(w_in) # avoid modifying global w_in
    # An array to store cost J and w's at each iteration primarily for graphing later
    J_history = []
    p_history = []
    b = b_in
    w = w_in
    
    for i in range(num_iters):
        # Calculate the gradient and update the parameters using gradient_function
        dj_dw, dj_db = gradient_function(x, y, w, b)     

        # Update Parameters using equation (3) above
        b = b - alpha * dj_db                            
        w = w - alpha * dj_dw                            

        # Save cost J at each iteration
        if i<100000:      # prevent resource exhaustion 
            J_history.append(cost_function(x, y, w, b))
            p_history.append([w,b])
        # Print cost every at intervals 10 times or as many iterations if < 10
        if i% math.ceil(num_iters/10) == 0:
            print(f"Iteration {i:4}: Cost {J_history[-1]:0.2e} ",
                  f"dj_dw: {dj_dw: 0.3e}, dj_db: {dj_db: 0.3e}  ",
                  f"w: {w: 0.3e}, b:{b: 0.5e}")
 
    return w, b, J_history, p_history #return w and J,w history for graphing

def compute_gradient(x, y, w, b): 
    """
    Computes the gradient for linear regression 
    Args:
      x (ndarray (m,)): Data, m examples 
      y (ndarray (m,)): target values
      w,b (scalar)    : model parameters  
    Returns
      dj_dw (scalar): The gradient of the cost w.r.t. the parameters w
      dj_db (scalar): The gradient of the cost w.r.t. the parameter b     
     """
    
    # Number of training examples
    m = x.shape[0]    
    dj_dw = 0
    dj_db = 0
    
    for i in range(m):  
        f_wb = w * x[i] + b 
        dj_dw_i = (f_wb - y[i]) * x[i] 
        dj_db_i = f_wb - y[i] 
        dj_db += dj_db_i
        dj_dw += dj_dw_i 
    dj_dw = dj_dw / m 
    dj_db = dj_db / m 
        
    return dj_dw, dj_db

#Function to calculate the cost
def compute_cost(x, y, w, b):
   
    m = x.shape[0] 
    cost = 0
    
    for i in range(m):
        f_wb = w * x[i] + b
        cost = cost + (f_wb - y[i])**2
    total_cost = 1 / (2 * m) * cost

    return total_cost

In [6]:
import math, copy
import numpy as np
import matplotlib.pyplot as plt

# Load our data set
x_train = np.array([1.0, 2.0])   #features
y_train = np.array([300.0, 500.0])   #target value

# initialize parameters
w_init = 0
b_init = 0
# some gradient descent settings
iterations = 10000
tmp_alpha = 1.0e-2
# run gradient descent
w_final, b_final, J_hist, p_hist = gradient_descent(x_train ,y_train, w_init, b_init, tmp_alpha, 
                                                    iterations, compute_cost, compute_gradient)
print(f"(w,b) found by gradient descent: ({w_final:8.4f},{b_final:8.4f})")
print(f"J_history: ({J_hist[-1]:5.3f})")

Iteration    0: Cost 7.93e+04  dj_dw: -6.500e+02, dj_db: -4.000e+02   w:  6.500e+00, b: 4.00000e+00
Iteration 1000: Cost 3.41e+00  dj_dw: -3.712e-01, dj_db:  6.007e-01   w:  1.949e+02, b: 1.08228e+02
Iteration 2000: Cost 7.93e-01  dj_dw: -1.789e-01, dj_db:  2.895e-01   w:  1.975e+02, b: 1.03966e+02
Iteration 3000: Cost 1.84e-01  dj_dw: -8.625e-02, dj_db:  1.396e-01   w:  1.988e+02, b: 1.01912e+02
Iteration 4000: Cost 4.28e-02  dj_dw: -4.158e-02, dj_db:  6.727e-02   w:  1.994e+02, b: 1.00922e+02
Iteration 5000: Cost 9.95e-03  dj_dw: -2.004e-02, dj_db:  3.243e-02   w:  1.997e+02, b: 1.00444e+02
Iteration 6000: Cost 2.31e-03  dj_dw: -9.660e-03, dj_db:  1.563e-02   w:  1.999e+02, b: 1.00214e+02
Iteration 7000: Cost 5.37e-04  dj_dw: -4.657e-03, dj_db:  7.535e-03   w:  1.999e+02, b: 1.00103e+02
Iteration 8000: Cost 1.25e-04  dj_dw: -2.245e-03, dj_db:  3.632e-03   w:  2.000e+02, b: 1.00050e+02
Iteration 9000: Cost 2.90e-05  dj_dw: -1.082e-03, dj_db:  1.751e-03   w:  2.000e+02, b: 1.00024e+02


### 我们用Linear regression为例：

__f-string 格式化__输出允许你在字符串中插入变量的值。具体来说，这行代码输出了一些关于迭代过程和梯度下降的信息，包括迭代次数、成本（损失）值、参数的梯度、以及参数 w 和 b 的值。

让我来解释一下代码中的不同部分：

f"Iteration {i:4}: Cost {J_history[-1]:0.2e} "：这个部分输出迭代次数和损失（成本）值。{i:4} 将迭代次数 i 格式化为占据 4 个字符的字符串(相当于小数点前保留)，使得输出的文本更整齐。{J_history[-1]:0.2e} 表示成本值，:0.2e 指定了科学计数法格式（指数为小数点后两位）。

f"dj_dw: {dj_dw: 0.3e}, dj_db: {dj_db: 0.3e} "：这个部分输出了梯度的值。{dj_dw: 0.3e} 和 {dj_db: 0.3e} 分别表示 dj_dw 和 dj_db 变量的值，使用科学计数法格式（指数为小数点后三位）。

f"w: {w: 0.3e}, b:{b: 0.5e}"：这个部分输出了参数 w 和 b 的值。{w: 0.3e} 和 {b: 0.5e} 分别表示 w 和 b 变量的值，同样使用科学计数法格式（指数为小数点后三位和五位）。

f"(w,b) found by gradient descent: ({w_final:8.4f},{b_final:8.4f})"：这个部分输出了最终参数 (w_final, b_final) 的值。{w_final:8.4f} 和 {b_final:8.4f} 分别表示 w_final 和 b_final 变量的值，使用格式 :8.4f，其中 8 表示字段宽度，. 后面的 4 表示小数点后的位数，f 表示浮点数格式。

整个代码行使用多个花括号 {} 来包裹变量，: 后面可以跟随一些格式选项，比如对齐、小数位数等。这样的格式化字符串使输出结果更易读，并且可以控制变量在输出中的呈现方式。

请注意，这段代码使用了 Python 3.6 之后引入的 f-string 特性。如果你的 Python 版本较低，可能需要使用旧的字符串格式化方式。